### PR TITLE
internal/v5: implement meta/published endpoint

### PR DESCRIPTION
**Pull request by rogpeppe**
This enables us to query any entity to find out what channels
it's in and whether it's the currently chosen entity for those channels.

### CODE REVIEW

File: dependencies.tsv [hunk 1]
```diff
@@ -13,6 +13,7 @@ github.com/juju/schema	git	1e25943f8c6fd6815282d6f1ac87091d21e14e19	2016-03-01T1
 github.com/juju/testing	git	ee18040b46bb1f8c93438383bd51ec77eb8c02ab	2016-01-12T21:04:04Z
 github.com/juju/txn	git	99ec629d0066a4d73c54d8e021a7fc1dc07df614	2015-06-09T16:58:27Z
 github.com/juju/utils	git	0cac78a34dd1c42d2f2dc718c345fd13e3a264fc	2016-01-29T15:50:19Z
+github.com/juju/version	git	102b12db83e38cb2ce7003544092ea7b0ca59e92	2015-11-07T04:32:11Z
```
Comment (frankban) [inline dependencies.tsv:16]: Why juju/version?

Comment (frankban) [inline dependencies.tsv:16]: Ah, maybe new resources stuff in charm or charmrepo.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1636/
Test PASSed.

File: internal/v5/api.go [hunk 2]
```diff
@@ -256,6 +247,16 @@ func RouterHandlers(h *ReqHandler) *router.Handlers {
 				h.putMetaCommonInfoWithKey,
 				"commoninfo",
 			),
+			"extra-info": h.puttableEntityHandler(
```
Comment (frankban) [inline internal/v5/api.go:250]: Thanks for reordering the endpoints.

File: internal/v4/relations_test.go [hunk 3]
```diff
@@ -226,6 +226,7 @@ var metaCharmRelatedTests = []struct {
 				Meta: map[string]interface{}{
 					"id-name": params.IdNameResponse{"wordpress"},
 					"charm-metadata": &charm.Meta{
+						Format: 1,
```
Comment (frankban) [inline internal/v4/relations_test.go:229]: What's this? Just curious.

File: internal/v5/api_test.go [hunk 4]
```diff
@@ -712,6 +733,136 @@ func (s *APISuite) TestMetaEndpointsSingle(c *gc.C) {
...
+	rev      int
+	channels []params.Channel
+}
+
+// Note that, unusually, all the entities in all the tests
+// are added before any of the "expect" values are
+// determined because we know that we want
+// exactly one test for each entity.
+var metaPublishedTests = []struct {
+	id       string
```
Comment (frankban) [inline internal/v5/api_test.go:746]: I think an about field would help future readers.

**Comment by frankban** [comment]
:+1:

**Comment by mhilton** [comment]
:+1:

File: internal/v4/relations_test.go (hunk 3 above)
Comment (rogpeppe) [inline internal/v4/relations_test.go:229]: It's a strange thing related to fixing yaml marshaling for charm.Meta. With the marshaling not fixed, the format field was getting written out explicitly (value 0), which meant that when we read the charm it did not acquire its default value (1 for some historical reason) but zero. Now that we're not writing the Format field (which is correct I think - Format is an ancient relic) we start seeing the usual default value of Format.

File: internal/v5/api_test.go (hunk 4 above)
Comment (rogpeppe) [inline internal/v5/api_test.go:746]: Given there's exactly one test per entity, I think we're probably OK here.

**Comment by rogpeppe** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Pull request merged**